kuma/spam/akismet.py

```diff
@@ -0,0 +1,174 @@
+import urlparse
+
+from constance import config
+from django.conf import settings
+from requests.packages.urllib3.util import Retry
+from requests.adapters import HTTPAdapter
+from requests import Session, exceptions
+
+
+class AkismetError(Exception):
+    """
+    A custom exception that takes the requests' response, its status code
+    and an optional debug help directly from Akismet.
+    """
+    def __init__(self, response, status_code, debug_help):
+        self.response = response
+        self.status_code = status_code
+        self.debug_help = debug_help
+
+    def __str__(self):
+        return ('%s response, debug help: %s, full response: %s' %
+                (self.status_code, self.debug_help, self.response.text))
+
+
+class Akismet(object):
+    """
+    The Akismet client implementation wrapping a request session and
+    having API specific methods for easy spam and ham handling.
+    """
+    submission_success = 'Thanks for making the web a better place.'
+
+    def __init__(self):
+        self.key = config.AKISMET_KEY
+        self.domain = settings.DOMAIN
+        self.session = Session()
+        self.adapter = HTTPAdapter(
+            max_retries=Retry(
+                # number of total retries
+                total=5,
+                # retry once in case we can't connect to Akismet
+                connect=1,
+                # retry once in case we can't read the response from Akismet
+                read=1,
+                # retry once in case we're redirect by Akismet
+                redirect=1,
+                # definitely retry if Akismet is unwell
+                status_forcelist=[500, 503])
+        )
+        self.session.mount('http://', self.adapter)
+        self.session.mount('https://', self.adapter)
+        # define if this client is even usable
+        # use this in the app code
+        try:
+            self.ready = self.verify_key()
+        except AkismetError:
+            # just take a shortcut when there was an
+            # error accessing Akismet
+            self.ready = False
+
+    @property
+    def url(self):
+        return 'https://%s.rest.akismet.com/1.1/' % self.key
+
+    def send(self, method, user_ip='', user_agent='', **payload):
+        payload.update({
+            'blog': self.domain,
+            'user_ip': user_ip,
+            'user_agent': user_ip,
+        })
+        url = urlparse.urljoin(self.url, method)
+        try:
+            return self.session.post(url, data=payload)
+        except exceptions.RequestException:
+            # let this exception bubble up for now since the mail_admins
+            # log handler and Sentry will let us know about this
+            raise
+
```

Comment:
I've thought about this a bit more and think it actually needs to pass without raising the exception since the premise is that content submission won't be hindered if there is a connectivity problem. Instead I'll record the exception with New Relic's agent and move on.